### ISSUE TITLE
Kobo eReader says "protected by Adobe DRM"

### ISSUE THREAD

**Issue by rgarrigue** [Michael-F-Bryan/mdbook-epub#28]
Hi there

I played with mdbook & this plugin yesterday, came up with https://github.com/rgarrigue/songbook. When I c/ced the resulting epub (you can DL it [here](https://github.com/rgarrigue/songbook/releases/tag/0.0.4)), my Kobo H2O reader says "Unauthorized to open this book due to Adobe DRM". I opened it successfully on my Ubuntu laptop with Okular. How can I fix this ? 

Best regards

**Comment by Michael-F-Bryan** [comment]
I'm surprised it thinks there's DRM involved. You may want to create a ticket against [lise-henry/epub-builder](https://github.com/lise-henry/epub-builder/) because that's the crate we use to generate the EPUB.

**Comment by justinclift** [comment]
As a general concept, could it be a case of "any file we can't open is probably something DRM protected"?

Saying that because we do similar with our [SQLite GUI project](https://github.com/sqlitebrowser/sqlitebrowser/).  We attempt to open a given database file, and when it doesn't work we assume it's encrypted and give a warning as such.

**Comment by JeremyAube** [comment]
https://github.com/Michael-F-Bryan/mdbook-epub/blob/640a81209b9de4af8da082222f31f62dc9d21af0/src/generator.rs#L28

I may have found something useful. EpubBuilder uses ZipLibrary to generate the epub, but the ZipLibrary object has a comment saying that it might cause an issue when exporting for ereaders.

From the ZipLibrary comment:
`While this has the advantage of not requiring an external zip command, I have run into some issues when trying to export EPUB generated with this method to ereaders (e.g. Kobo).`

**Comment by JeremyAube** [comment]
After a little bit of testing I can confirm that the ZipLibrary is actually causing the error.

## Temporary fix
1. run `unzip <book-name>.epub` to unzip the book;
2. run `rm <book-name>.epub`
3. run `zip -rX <book-name>.epub mimetype META-INF OEBPS` to rezip the book;

you can then transfer the new book to your kobo and it should work as expected.

**Comment by Michael-F-Bryan** [comment]
The proposed suggestion of re-zipping the generated document aren't overly satisfying, 

I think the best way to solve the problem would be to look at the `ZipLibrary` implementation, figure out what it's doing to make the Kobo eReader think it's protected by DRM, then remove that behaviour. `epub-builder` is free ([as in speech](https://github.com/lise-henry/epub-builder/#license)) software and doesn't add any Adobe DRM to the book.

**Comment by rgarrigue** [comment]
On the last point I'll quote @lise-henry hint on the issue I opened on its library repo 

> From what I see, the HTML files only contain the body part and is missing the beginning and end for a full XHTML that is required to be understood as valid, e.g.

```
<?xml version="1.0" encoding="UTF-8"?>
<html xmlns="http://www.w3.org/1999/xhtml" xmlns:epub="http://www.idpf.org/2007/ops">
<body>
...
</body>
</html>
```

So I guess the Kobo misses this and a generic error message says DRM. If you want to confirm, you can share some test epub that'll test on my Kobo.

**Comment by justinclift** [comment]
Would it be feasible to use https://github.com/mvdnes/zip-rs instead?

**Comment by JeremyAube** [comment]
@Michael-F-Bryan 
I do think that this is the best solution for the future but in the meantime we can fallback to command line if it is available on the system. I don't think the `zip` crate will be updated anytime soon as it requires a lot of digging and the Kobo error messages are not helpful. I will do some tests to validate that the issue is in fact with the `zip` crate and open an issue if it is.

@rgarrigue 
Although anything is possible, I don't believe that it necessarily causes a problem. If you take a generated book and unzip/rezip with the system's zip command (ubuntu 20.04 in my case), the book will work on Kobo without any modifications to the XHTML files.

@justinclift 
https://github.com/mvdness/zip-rs might actually be the problem. `ZipLibrary` seems to just be a wrapper around that library.

**Comment by justinclift** [comment]
> https://github.com/mvdness/zip-rs might actually be the problem. ZipLibrary seems to just be a wrapper around that library.

Ahhh.  If they're open to PR's, it might be a simple fix. :smile:

**Comment by JeremyAube** [comment]
I will first validate that `zip-rs` is actually the problem and then we'll see. Do you have a suggestion on a replacement?

**Comment by JeremyAube** [comment]
The issue is now fixed in zip-rs. Once updated in Epub-Builder library we should update the dependency here as well.

**Comment by JeremyAube** [comment]
https://github.com/lise-henry/epub-builder/issues/11

epub-builder got updated and fixed this issue. we just need a version a version bumb and everything should work now!

**Comment by crone123** [comment]
Hello,
Is this issue normaly fixed?
I tired yesterday to mdbook-epub the rust book, rust async book, and rust nomicon, and the 3 files are recognizer as "Protected by Adobe DRM" on my Kobo Libra H2O (with latest updates).

If it's fixed, why the version I got with cargo install have this bug? How to install a fixed version?

Thank you !
Valentin CRÔNE

**Comment by Michael-F-Bryan** [comment]
@crone123, see the first checkbox in #34.

**Comment by blandger** [comment]
Most probably the 'DRM error' is caused by one song title:
Simon & Garfunkel
because symbols like **& < >** must be replaced by **& amp ; & lt ;  & gt ;** inside several epub generated files (used for XHTML navigation files)
See my PR = https://github.com/lise-henry/epub-builder/pull/15
into epub-builder library.

**Comment by blandger** [comment]
@crone123 
Can you take any released book from repo
https://github.com/rust-lang-ru/book/releases/tag/0.0.1
and check it on your reader ?

Those epub books are generated by my custom mdbook-epub version with all necessary fixes.

**Comment by crone123** [comment]
Hello,
The book works perfectly with my reader (Kobo Libra H2O), I can preview and open it, the table of contents works, this is very good !
I hope this fix will be merged soon in the crates.io version :)
Thank you for you help ! :D

**Comment by blandger** [comment]
should be fixed by new 0.5.0 version
https://docs.rs/epub-builder/0.5.0/epub_builder/
